File: backend/src/routes/network.ts
import { Router, Request, Response } from 'express';
import { db } from '../database/db.js';

const router = Router();

// GET /partners
router.get('/partners', (req: Request, res: Response) => {
  const category = req.query.category as string;
  let partners = db.getCollection("partners");

  if (category && category !== 'SEMUA') {
    partners = partners.filter((p: any) => p.category === category);
  }
  res.json(partners);
});

// POST /connect/:partnerId
router.post('/connect/:partnerId', (req: Request, res: Response) => {
  const { partnerId } = req.params;
  const partners = db.getCollection("partners");
  const partner = partners.find((p: any) => p.id === partnerId);

  if (!partner) {
    return res.status(404).json({ detail: "Mitra tidak ditemukan" });
  }

  const newStatus = !partner.is_connected;
  db.update("partners", partnerId, { is_connected: newStatus });

  res.json({
    is_connected: newStatus,
    message: newStatus
      ? `Terhubung dengan ${partner.name}. Anda dapat langsung menawarkan hasil panen.`
      : `Koneksi dengan ${partner.name} diputuskan.`
  });
});

// POST /offer-harvest
router.post('/offer-harvest', (req: Request, res: Response) => {
  const payload = req.body;
  const partners = db.getCollection("partners");
  const partner = partners.find((p: any) => p.id === payload.partner_id);

  if (!partner) {
    return res.status(404).json({ detail: "Mitra tujuan tidak ditemukan" });
  }

  const priceText = payload.offered_price_per_kg
    ? `Rp ${payload.offered_price_per_kg.toLocaleString('id-ID')}/kg`
    : "Sesuai harga pasar terbaik";

  const message = `Halo ${partner.pic_name}, saya dari AgriBuddy ingin menawarkan hasil panen:\n` +
    `🌾 Komoditas: ${payload.commodity}\n` +
    `⚖️ Estimasi Volume: ${payload.weight_kg} kg\n` +
    `💰 Harapan Harga: ${priceText}\n` +
    (payload.notes ? `📝 Catatan: ${payload.notes}\n` : '') +
    `Apakah KUD/Penggilingan siap menyerap panen kami? Terima kasih.`;

  const phone = partner.phone_whatsapp.replace(/\D/g, '');
  const cleanPhone = phone.startsWith('0') ? '62' + phone.substring(1) : phone;
  const waUrl = `https://wa.me/${cleanPhone}?text=${encodeURIComponent(message)}`;

  const now = new Date();
  const options: Intl.DateTimeFormatOptions = {
    day: '2-digit',
    month: 'short',
    year: 'numeric',
    hour: '2-digit',
    minute: '2-digit',
    hour12: false
  };
  const createdAt = `${now.toLocaleDateString('id-ID', options)} WIB`;

  const newOffer = {
    partner_id: partner.id,
    partner_name: partner.name,
    commodity: payload.commodity,
    weight_kg: payload.weight_kg,
    offered_price_per_kg: payload.offered_price_per_kg,
    status: "MENUNGGU_RESPON",
    created_at: createdAt,
    whatsapp_url: waUrl
  };

  const inserted = db.insert("offers", newOffer);
  res.json(inserted);
});

// GET /offers
router.get('/offers', (req: Request, res: Response) => {
  const offers = db.getCollection("offers");
  res.json(offers);
});

export default router;

File: backend/legacy_python/app/api/network.py
import urllib.parse
from datetime import datetime
from typing import List, Optional, Dict, Any
from fastapi import APIRouter, HTTPException, Query
from app.models.schemas import PartnerProfile, HarvestOfferCreate, HarvestOfferResponse
from app.core.database import db

router = APIRouter(prefix="/network", tags=["Jejaring Sosial & Distribusi Pertanian"])

@router.get("/partners", response_model=List[Dict[str, Any]])
def get_partners(category: Optional[str] = Query(None, description="Filter: KOPERASI, DISTRIBUTOR_PUPUK, PENGGILINGAN_PASAR")):
    partners = db.get_collection("partners")
    if category and category != "SEMUA":
        return [p for p in partners if p.get("category") == category]
    return partners

@router.post("/connect/{partner_id}")
def toggle_partner_connection(partner_id: str):
    partners = db.get_collection("partners")
    partner = next((p for p in partners if p.get("id") == partner_id), None)
    if not partner:
        raise HTTPException(status_code=404, detail="Mitra tidak ditemukan")
    
    new_status = not partner.get("is_connected", False)
    updated = db.update("partners", partner_id, {"is_connected": new_status})
    return {
        "partner_id": partner_id,
        "is_connected": new_status,
        "message": "Berhasil terhubung dengan mitra" if new_status else "Koneksi kemitraan dilepas"
    }

@router.post("/offer-harvest", response_model=HarvestOfferResponse)
def create_harvest_offer(payload: HarvestOfferCreate):
    partners = db.get_collection("partners")
    partner = next((p for p in partners if p.get("id") == payload.partner_id), None)
    if not partner:
        raise HTTPException(status_code=404, detail="Mitra tujuan tidak ditemukan")
    
    # Generate WhatsApp Deep-Link dengan pesan profesional ramah petani
    price_text = f"Rp {payload.offered_price_per_kg:,.0f}/kg" if payload.offered_price_per_kg else "Harga negosiasi terbaik"
    wa_message = (
        f"Halo {partner['name']},\n"
        f"Saya Pak Joko (Petani Binaan AgriBuddy - Desa Sukamaju).\n"
        f"Saya memiliki stok hasil panen di lumbung:\n"
        f"- Komoditas: {payload.commodity}\n"
        f"- Kuantitas: {payload.weight_kg:,.0f} Kg\n"
        f"- Harapan Harga: {price_text}\n"
        f"- Catatan: {payload.notes or 'Siap angkut'}\n\n"
        f"Apakah pihak {partner['name']} berminat untuk penyerapan komoditas ini? Terima kasih."
    )
    encoded_message = urllib.parse.quote(wa_message)
    wa_url = f"https://wa.me/{partner['phone_whatsapp']}?text={encoded_message}"
    
    new_offer = {
        "partner_id": payload.partner_id,
        "partner_name": partner["name"],
        "commodity": payload.commodity,
        "weight_kg": payload.weight_kg,
        "offered_price_per_kg": payload.offered_price_per_kg,
        "status": "DITAWARKAN",
        "created_at": datetime.now().strftime("%d %b %Y, %H:%M WIB"),
        "whatsapp_url": wa_url,
        "notes": payload.notes
    }
    inserted = db.insert("offers", new_offer)
    return HarvestOfferResponse(
        id=inserted["id"],
        partner_name=inserted["partner_name"],
        commodity=inserted["commodity"],
        weight_kg=inserted["weight_kg"],
        offered_price_per_kg=inserted.get("offered_price_per_kg"),
        status=inserted["status"],
        created_at=inserted["created_at"],
        whatsapp_url=inserted["whatsapp_url"]
    )

@router.get("/offers", response_model=List[Dict[str, Any]])
def get_my_offers():
    return db.get_collection("offers")
